### PR TITLE
Serialize game state into json

### PR DESCRIPTION
**Pull request by mplantady**
Serialize game state into human readable JSON.
The save is gametype/agent agnostic, to be able to reload a board with a different setup.

**Record game example:**
python run_battle.py --agents=test::a.pommerman.agents.SimpleAgent,test::a.pommerman.agents.SimpleAgent,test::a.pommerman.agents.SimpleAgent,test::a.pommerman.agents.SimpleAgent --config=ffa_v0 --record_json_dir=./jsons

**Replay game from step 50 example:**
python run_battle.py --agents=test::a.pommerman.agents.SimpleAgent,test::a.pommerman.agents.SimpleAgent,test::a.pommerman.agents.SimpleAgent,test::a.pommerman.agents.SimpleAgent --config=ffa_v0 --game_state_file=./jsons/50.json

### CODE REVIEW

File: games/a/pommerman/envs/utility.py [hunk 1]
```diff
@@ -317,12 +319,18 @@ def get_next_position(position, direction):
...
+    return json.dumps(obj, default=serializer)
+
+def serializer(obj):
+    if isinstance(obj, numpy.ndarray):
+        return obj.tolist()
+    elif isinstance(obj, numpy.int64):
+        return int(obj)
+    elif hasattr(obj, 'to_json'):
+        return obj.to_json()
+    elif hasattr(obj, '__dict__'):
```
Comment (cinjon) [inline games/a/pommerman/envs/utility.py]: should this be recursive?

File: games/a/pommerman/envs/v0.py [hunk 2]
```diff
@@ -433,11 +433,9 @@ def render(self, mode='human', close=False, record_pngs_dir=None, record_json_di
                 self._viewer = None
             return
 
-        if record_pngs_dir:
-            assert not os.path.isdir(record_pngs_dir)
+        if record_pngs_dir and not os.path.isdir(record_pngs_dir):
             os.makedirs(record_pngs_dir)
-        if record_json_dir:
-            assert not os.path.isdir(record_json_dir)
+        if record_json_dir and not os.path.isdir(record_json_dir):
```
Comment (cinjon) [inline games/a/pommerman/envs/v0.py]: if the directories already exist, then this is going to silently keep writing to them[, and possibly overriding]. that seems bad.

File: games/a/pommerman/envs/v0.py [hunk 3]
```diff
@@ -463,8 +461,8 @@ def render(self, mode='human', close=False, record_pngs_dir=None, record_json_di
             Image.fromarray(img).save(os.path.join(record_pngs_dir, '%d.png' % self._step_count))
         if record_json_dir:
             info = self.get_json_info()
-            with open(os.path.join(record_json_dir, '%d.png' % self._step_count), 'w') as f:
-                f.write(json.dumps(info))
+            with open(os.path.join(record_json_dir, '%d.json' % self._step_count), 'w') as f:
+                f.write(json.dumps(info, sort_keys=True, indent=4))
```
Comment (cinjon) [inline games/a/pommerman/envs/v0.py]: nice.

File: games/a/pommerman/envs/v0.py [hunk 4]
```diff
@@ -501,20 +499,44 @@ def featurize(obs):
...
+        self._step_count = self._init_game_state['step_count']
+        self._board_size = self._init_game_state['board_size']
+
+        boardArray = json.loads(self._init_game_state['board'])
+        self._board = np.ones((self._board_size, self._board_size)).astype(np.uint8) * utility.Item.Passage.value
+        for x in range(self._board_size):
+            for y in range(self._board_size):
+                self._board[x,y] = boardArray[x][y]
+
+        self._items = eval(json.loads(self._init_game_state['items']))
```
Comment (cinjon) [inline games/a/pommerman/envs/v0.py]: why is this eval?

**Comment by cinjon** [comment]
Nice. This is great. Added some small comments in there, otherwise looking solid.

**Comment by cinjon** [comment]
Helpful to have a brief comment specifying what ```def set_init_game_state(self, game_state_file)``` expects in the game_state_file.

File: games/a/pommerman/envs/v0.py (hunk 4 above)
Comment (mplantady) [reply games/a/pommerman/envs/v0.py]: Dictionary with int key tuple don't serialize well. I wrap the all 'items' section into a string and reverse it with eval, but with hindsight sounds quite hackish

File: games/a/pommerman/envs/utility.py (hunk 1 above)
Comment (mplantady) [reply games/a/pommerman/envs/utility.py]: I believe the default serialization method is called recursively on nested object

**Pull request merged**